File: programmers/high_score_kit/완전탐색/카펫.py
def solution(brown, yellow):
    answer = []
    numbers = list()
    numCarpet = brown + yellow
    for num in range(1, numCarpet + 1):
        if (numCarpet) % num == 0:
            numbers.append(num)

    for num in numbers:
        if numCarpet // num in numbers and numCarpet // num <= num:
            answer = [num, numCarpet // num]
            return answer

print(solution(10, 2)) #[4, 3]
print(solution(8, 1)) #[3, 3]
print(solution(24, 24)) #[8, 6]
print(solution(14, 4)) #[6, 3]

File: programmers/high_score_kit/완전탐색/카펫.js
const permutation = (arr, selectNum) => {
    const result = [];
    if (selectNum === 1) return arr.map((v) => [v]);

    arr.forEach((v, idx, arr) => {
        const fixed = v;
        const restArr = arr;
        const permutationArr = permutation(restArr, selectNum - 1);
        const combineFix = permutationArr.map((v) => [fixed, ...v]);
        result.push(...combineFix);
    });
    return result;
}

function solution(brown, yellow) {
    let answer = [];
    const sum = brown + yellow;
    let yaksu = [];
    for (let i = 1; i < sum + 1; i++) {
        if (sum % i === 0) {
            yaksu.push(i);
        }
    }
    const permuList = permutation(yaksu, 2);
    for (let i = 0; i < permuList.length; i++) {
        const [w, h] = [permuList[i][0], permuList[i][1]];
        if (w >= h && w * h === sum && (w - 2) * (h - 2) === yellow) {
            answer = [w, h];
        }
    }
    return answer;
}

console.log(solution(10, 2)); // [4, 3]